core/snmp/impl-snmp4j/src/main/java/org/opennms/netmgt/snmp/snmp4j/opaqueadapter/StringWithNumberCheckAdapter.java

```diff
@@ -0,0 +1,86 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2022 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2022 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.snmp.snmp4j.opaqueadapter;
+
+import java.util.Objects;
+import org.opennms.netmgt.snmp.snmp4j.OpaqueValueType;
+
+
+public class StringWithNumberCheckAdapter implements OpaqueTypeAdapter {
+    
+    String stringValue;
+    Long longValue = null;
+    Double doubleValue = null;
+    OpaqueValueType valueType = OpaqueValueType.STRING;
+
+    public StringWithNumberCheckAdapter(String value) {
+        Objects.requireNonNull(value);
+        this.stringValue = value;
+        
+        //Try to extract a long value
+        try {
+            this.longValue = Long.parseLong(value);
+            this.doubleValue = (double)this.longValue;
+            this.valueType = OpaqueValueType.LONG;
+            return;
+        } catch (NumberFormatException ex) {
+            //do nothing
+        }
+        
+        //Try to extract a double value
+        try {
+            this.doubleValue = Double.valueOf(value);
+            this.longValue = this.doubleValue.longValue();
+            this.valueType = OpaqueValueType.DOUBLE;
+        } catch (NumberFormatException ex) {
+            //do nothing
+        }
```

Comment:
This detection logic is rely on NumberFormatException? Is there any better way to do it? It is kind of difficult to understand.

Reply:
When a number should be transferred as String wrapped into Opaque Type then it must be extracted out of it.
To prevent multiple conversions each time in code like [ if (x.isNumric()) y=x.getLong() ].... it is converted  just one time if possible or it is not a number but just a string.
When it is a String "123" then it will be parsed as long, when "123.4e45"  then as double. And later it can be stored into rrd.
In fact I have just implemented described in Ticket format. (switch (type) { case 120:.......) But in ticket is mentioned it could be a string. ( case ???: return StringWithNumberCheckAdapter(....<extract encoded string value here>...) ) 
When for some snmp data pure strings are needed, then PureStringAdapter without "time-consuming"  conversion can also be easily implemented.

Reply:
If that is the case you at least need to leave more comment to explain your try catch logic. And why the double should override long in the second try. Or may be you can test double first. And only try long if double fail ? I am not sure just by guessing.

Reply:
Ok. I'll describe it in comment